### PR TITLE
Add GraphQL interface with ballerina distinct object type

### PR DESCRIPTION
**Pull request by MohamedSabthar**
## Purpose

Fixes: [#3289](https://github.com/ballerina-platform/ballerina-standard-library/issues/3289)

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec

### CODE REVIEW

**Comment by MohamedSabthar** [comment]
PR depends on fixes indroduced in `v2201.2.0-stage` branch of lang repo.
For the feature to work ballerinaLangVersion should be updated after new release

File: changelog.md [hunk 1]
```diff
@@ -5,11 +5,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+- [[#3289] Support GraphQL interface with ballerina distinct object type](https://github.com/ballerina-platform/ballerina-standard-library/issues/3289)
 
 ### Added
 - [[#2898] Support Deprecation Support in GraphQL Services](https://github.com/ballerina-platform/ballerina-standard-library/issues/2898)
 - [[#2001] Support GraphQL Interceptors for Query and Mutation operations](https://github.com/ballerina-platform/ballerina-standard-library/issues/2001)
 - [[#3260] Log the Errors Returned from the Resolvers](https://github.com/ballerina-platform/ballerina-standard-library/issues/3260)
+- [[#3289] Support GraphQL interface with ballerina distinct object type](https://github.com/ballerina-platform/ballerina-standard-library/issues/3289)
```
Comment (DimuthuMadushan) [inline changelog.md:14]:
```suggestion
```

File: changelog.md [hunk 2]
```diff
@@ -5,11 +5,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+- [[#3289] Support GraphQL interface with ballerina distinct object type](https://github.com/ballerina-platform/ballerina-standard-library/issues/3289)
```
Comment (DimuthuMadushan) [inline changelog.md:8]:
```suggestion
## [Unreleased]

### Added
- [[#3289] Support GraphQL interface with ballerina distinct object type](https://github.com/ballerina-platform/ballerina-standard-library/issues/3289)
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java [hunk 3]
```diff
@@ -141,7 +162,7 @@ private Field getField(ResourceMethodSymbol methodSymbol, List<PathSegment> list
             String deprecationReason = getDeprecationReason(methodSymbol);
             Type fieldType = getType(methodSymbol);
             Field field = new Field(list.get(0).signature(), getDescription(methodSymbol), fieldType, isDeprecated,
-                                    deprecationReason);
+                    deprecationReason);
```
Comment (DimuthuMadushan) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:165]: Let's keep this as earlier. check other places as well.

Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java]: Fixed

File: docs/spec/spec.md [hunk 4]
```diff
@@ -766,7 +758,7 @@ public isolated distinct service class Teacher {
 }
 ```
 
-In the above example, the `Human` class is an interface. The `Student` and `Teacher` classes are `Object` types that implement the `Human` interface.
+In the above example, the `Person` class is an interface. The `Student` and `Teacher` classes are `Object` types that implement the `Person` interface.
```
Comment (DimuthuMadushan) [inline docs/spec/spec.md:761]:
```suggestion
In the above example, the `Person` object is an interface. The `Student` and `Teacher` classes are `Object` types that implement the `Person` interface.
```

**Review by DimuthuMadushan — APPROVED** [review]

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java [hunk 5]
```diff
@@ -93,6 +94,26 @@ public SchemaGenerator(ServiceDeclarationSymbol serviceDeclarationSymbol, Interf
...
+        return false;
+    }
+
+    private static String getParameterDescription(String parameterName, MethodSymbol methodSymbol) {
+        if (methodSymbol.documentation().isEmpty()) {
+            return null;
+        }
+        return methodSymbol.documentation().get().parameterMap().get(parameterName);
+    }
+
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:116]: These are automatically rearranged by the IDE

File: changelog.md [hunk 6]
```diff
@@ -23,6 +24,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - [[#3115] Fix Not Initializing Context per Request in GraphQL Subscriptions](https://github.com/ballerina-platform/ballerina-standard-library/issues/3115)
 - [[#3227] Fix Service Crashing when Map Field Does not Contain the Provided Key](https://github.com/ballerina-platform/ballerina-standard-library/issues/3227)
 - [[#2897] Fix GraphiQL Schema not Generating for Default Values](https://github.com/ballerina-platform/ballerina-standard-library/issues/2897)
+- [[#2897] Fix Invalid introspection response for fields with default value](https://github.com/ballerina-platform/ballerina-standard-library/issues/3307)
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply changelog.md:27]: Yes this is fixed in 1.4.0, changelog updated after noticing we have missed this to update in the log

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java [hunk 7]
```diff
@@ -294,19 +302,30 @@ private Type getType(String name, TypeDefinitionSymbol typeDefinitionSymbol) {
             return getType(name, description, parameterMap, intersectionType);
         } else if (typeDefinitionSymbol.typeDescriptor().typeKind() == TypeDescKind.TABLE) {
             return getType((TableTypeSymbol) typeDefinitionSymbol.typeDescriptor());
+        } else if (typeDefinitionSymbol.typeDescriptor().typeKind() == TypeDescKind.OBJECT) {
+            return getTypeFromObjectDefinition(name, typeDefinitionSymbol);
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:306]: this method is already inside the `getType(String name, TypeDefinitionSymbol typeDefinitionSymbol)` signature. We already have a overloaded method with the same signature

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java [hunk 8]
```diff
@@ -316,11 +335,20 @@ private Type getType(String name, ClassSymbol classSymbol) {
...
             // When adding an implementation, the name is already checked. Therefore, no need to check isEmpty().
-            //noinspection OptionalGetWithoutIsPresent
-            Type implementedType = getType(implementation.getName().get(), implementation);
+            // noinspection OptionalGetWithoutIsPresent
+            String implementationName = implementation.getName().get();
+            if (implementation.kind() == SymbolKind.CLASS) {
+                implementedType = getType(implementationName, (ClassSymbol) implementation);
+            } else {
+                implementedType = getTypeFromObjectDefinition(implementationName,
+                                                              (TypeDefinitionSymbol) implementation);
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:349]: the implementation of an interface an only either be class or type definition of an object. SymbolKind check is not needed here

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/Utils.java [hunk 9]
```diff
@@ -120,19 +121,22 @@ public static TypeSymbol getEffectiveType(IntersectionTypeSymbol intersectionTyp
         return intersectionTypeSymbol;
     }
 
-    public static boolean isDistinctServiceReference(TypeSymbol typeSymbol) {
-        if (typeSymbol.typeKind() != TypeDescKind.TYPE_REFERENCE) {
-            return false;
+    public static ObjectTypeSymbol getObjectTypeSymbol(Symbol serviceObjectTypeOrClass) {
+        if (serviceObjectTypeOrClass.kind() == SymbolKind.TYPE_DEFINITION) {
+            TypeDefinitionSymbol serviceObjectTypeSymbol = (TypeDefinitionSymbol) serviceObjectTypeOrClass;
+            return (ObjectTypeSymbol) serviceObjectTypeSymbol.typeDescriptor();
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/Utils.java:127]: Fixed in : [#919](https://github.com/ballerina-platform/module-ballerina-graphql/pull/919)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/service/validator/ServiceValidator.java [hunk 10]
```diff
@@ -329,51 +329,80 @@ private void validateReturnTypeClass(ClassSymbol classSymbol, Location location)
...
+        validateInterfaceObjectTypeDefinition(typeDefinitionSymbol, location);
+        validateInterfaceImplementation(objectTypeName, location);
+    }
+
+    private void validateInterfaceObjectTypeDefinition(TypeDefinitionSymbol typeDefinitionSymbol, Location location) {
+        if (this.visitedClassesAndObjectTypeDefinitions.contains(typeDefinitionSymbol)) {
+            return;
+        }
+        this.visitedClassesAndObjectTypeDefinitions.add(typeDefinitionSymbol);
+        // TODO: Check for distinct keyword and add diagnostic
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/service/validator/ServiceValidator.java:371]: Updated in : [#919](https://github.com/ballerina-platform/module-ballerina-graphql/pull/919)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:306]: In that case, we should use the same method. No point in defining two methods to do the same thing, right?
I guess in this case, what we need is to define a new method `getType(String name, ObjectTypeSymbol objectTypeSymbol, String description)`.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:116]: Shall we keep them as they were before? Additionally, can you check your IDE settings to refrain from doing such changes in the future?

[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:116]: Updated in #919

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/SchemaGenerator.java:306]: Updated in #919

**Pull request merged**